baker/GrapherBaker.tsx

```diff
@@ -134,6 +133,8 @@ const deleteOldGraphers = async (bakedSiteDir: string, newSlugs: string[]) => {
             slug.replace(`${bakedSiteDir}/grapher/`, "").replace(".html", "")
         )
     const toRemove = without(oldSlugs, ...newSlugs)
+        // do not delete grapher slugs redirected to explorers
+        .filter((slug) => !isPathRedirectedToExplorer(`/grapher/${slug}`))
```

Comment:
I'm missing a few things here:

- why are we keeping the grapher html and png export around when that grapher is being redirected to an explorer?
- does this assume that an explorer redirect is created before the grapher chart is deleted? What happens in the unlikely opposite case or if we were to bake from scratch, without an existing bakedSiteDir?
- should grapher -> explorer redirects defined in Wordpress be considered?

Reply:
All great questions, this is just adding another layer of duct tape. And I didn't explain this PR well: what currently happens is Grapher → Explorer redirects get baked, but are then erased by this step, which is why I added a skip condition here.

> why are we keeping the grapher html and png export around when that grapher is being redirected to an explorer?

**The HTML gets overwritten in a step before**. These Grapher → Explorer redirects get applied client-side in order to pass on query parameters, since as you pointed out, query parameters get dropped by Netlify if the target URL already contains query parameters. So a page gets baked at `/grapher/some-chart-slug` and it uses `history.replaceState` to change the URL on the client-side.
  
   **The PNG & SVG exports should be deleted**, I missed to account for that. But there are two cases:
      
   1. The previous bake had a grapher which in the new bake is replaced with a redirect. In this case we want to delete PNG & SVG and replace the HTML.
   2. The previous bake and the current bake both have an explorer redirect. Then, the PNG & SVG wouldn't exist, so we shouldn't attempt to delete them. Currently, the baker deletes HTML, SVG and PNG. The HTML is generated by the Grapher → Explorer redirect, this is the reason we get a whole bunch of  we get a whole bunch of DELETE errors at the end of every deploy, which is 

   Anyway, it's messy, there is probably a better way to handle this, but it probably needs quite a bit of thinking. The core issue is that the client-side explorer redirects bake files in the same directory as normal charts, so it leads to a kind of namespace clash. The thing is, these Grapher → Explorer redirects can be applied to URLs other than graphers, this is how `/coronavirus-data-explorer` → `/explorers/coronavirus-data-explorer` redirect is implemented too.

> does this assume that an explorer redirect is created before the grapher chart is deleted?

From an authoring point of view, no, it doesn't matter whether you add a redirect first or delete a chart.

From a baking point of view, I think also no, because we are just skipping deletion in this case, which is the cause for problems currently: because the explorer redirects get baked first, this step deletes them when it shouldn't.

> should grapher -> explorer redirects defined in Wordpress be considered?

Interesting question! I was thinking that it would be nice for authors to be able to add redirects, so they are not, as currently, only controlled by devs: https://github.com/owid/owid-grapher/blob/6a5729d34556a241bbc38b0c1cf72e105c6af3a2/explorerAdminServer/ExplorerRedirects.ts#L4-L201

Potentially we could handle Grapher → Explorer redirects better by omitting them from Netlify `_redirects` and instead implementing these client-side redirects. One slight issue is that the way they're currently architected, they need a `ExplorerUrlMigrationId` identifier associated with each redirect, so we know how to handle them: 

https://github.com/owid/owid-grapher/blob/6a5729d34556a241bbc38b0c1cf72e105c6af3a2/explorer/urlMigrations/ExplorerUrlMigrations.ts#L11-L13

https://github.com/owid/owid-grapher/blob/6a5729d34556a241bbc38b0c1cf72e105c6af3a2/explorer/urlMigrations/LegacyCovidUrlMigration.ts#L111-L131

Reply:
thanks for the very detailed explanation, I appreciate it!

> The HTML gets overwritten in a step before. These Grapher → Explorer redirects get applied client-side

That was the missing ingredient for me to understand, thanks for pointing it out. Explorer redirects [get baked in SiteBaker as HTML files with client-side redirects](https://github.com/owid/owid-grapher/blob/fix-explorer-redirects/baker/SiteBaker.tsx#L232) (as opposed to being added to _redirects), then **not** deleted in GrapherBaker.

> what currently happens is Grapher → Explorer redirects get baked, but are then erased by this step

I think I have a similar sense of "too many cooks" when looking [at the exclusions in the SiteBaker](https://github.com/owid/owid-grapher/blob/d4c3d1724e9c00276bd82cb534cd6718f64010cd/baker/SiteBaker.tsx#L146), where we need to make sure that function A doesn't delete legitimate paths that have been baked by function B. The other side of that same coin - name clashes - is equally scary.